### PR TITLE
Use a PlayerInfo transaction to prevent saving state from mid-conversation

### PR DESCRIPTION
**Pull request by wjp**
**Bugfix:** This PR addresses issue #3746

## Fix Details
The approach taken is pre-emptively saving the PlayerInfo to an in-memory savegame, and using that state when saving during a transaction. This should be a robust way of handling this without adding a lot of infrastructure to keep track of two states.

This patch modifies DataWriter to make it possible to keep a temporary state.

Earlier commits in this branch have debugging info to trace when these functions get called, including timing to show how long taking the snapshot takes (not long).

## Testing Done
Tested with saving/loading during spaceport missions and missions triggering during boarding. I have also added a first attempt at an integration test for this that fails on master and passes on this branch.

## Save File
N/A

### CODE REVIEW

File: source/DataWriter.cpp [hunk 1]
```diff
@@ -40,10 +40,16 @@ DataWriter::DataWriter(const string &path)
 // Destructor, which saves the file all in one block.
 DataWriter::~DataWriter()
 {
-	Files::Write(path, out.str());
+	if(!path.empty())
+		SaveToPath(path);
 }
```
Comment (RisingLeaf) [inline source/DataWriter.cpp:46]:
```suggestion


```

File: source/DataWriter.cpp [hunk 2]
```diff
@@ -40,10 +40,16 @@ DataWriter::DataWriter(const string &path)
...
 
 
+void DataWriter::SaveToPath(const std::string &filepath)
+{
+	Files::Write(filepath, out.str());
+}
+
```
Comment (RisingLeaf) [inline source/DataWriter.cpp:53]:
```suggestion



```

File: source/DataWriter.h [hunk 3]
```diff
@@ -34,6 +34,7 @@ class DataNode;
 class DataWriter {
 public:
 	// Constructor, specifying the file to write.
+	// An empty path will create a DataWriter that will not save its contents.
```
Comment (RisingLeaf) [inline source/DataWriter.h:37]: I would prefer another constructor that just takes no path

File: source/PlayerInfo.cpp [hunk 4]
```diff
@@ -3748,6 +3788,10 @@ void PlayerInfo::Save(const string &path) const
 			out.Write(plugin.first);
 		out.EndChild();
 	}
+
+	const auto stop = chrono::high_resolution_clock::now();
+	fprintf(stderr, "Saved in %lld us\n",
```
Comment (RisingLeaf) [inline source/PlayerInfo.cpp:3793]: is this part of your debuging?

**Review by RisingLeaf — CHANGES_REQUESTED** [review]

File: source/PlayerInfo.cpp (hunk 4 above)
Comment (wjp) [reply source/PlayerInfo.cpp:3793]: Yes indeed; I wanted to make sure the often unnecessary serialization didn't have a noticable performance impact.

Comment (RisingLeaf) [reply source/PlayerInfo.cpp:3793]: you might want to remove it though

**Comment by RisingLeaf** [comment]
Oh, please resolve the review comments and maybe say something like "done", so I get noticed about it

**Review by RisingLeaf — APPROVED** [review]
Structural approval earned

File: source/DataWriter.h (hunk 3 above)
Comment (wjp) [reply source/DataWriter.h]: Done

File: source/PlayerInfo.cpp (hunk 4 above)
Comment (wjp) [reply source/PlayerInfo.cpp]: Certainly

**Comment by wjp** [comment]
Ah, got it. Still getting used to github's review system.

Thanks for the review and comments!

**Comment by wjp** [comment]
I have done some more testing in the meantime, with spaceport missions and missions triggering during boarding, and both seem to behave the way I'm expecting. I have now also added a first attempt at an integration test for this that fails on master and passes on this branch.

This is the first time I'm writing any "content"/missions so I hope this test does what I think it does.

**Review by UnorderedSigh — APPROVED** [review]
It works for me, with both landing and spaceport missions.

I used this file, which gifts ships, and I removed "landing" to test spaceport:

[test.txt](https://github.com/endless-sky/endless-sky/files/10214373/test.txt)

**Comment by UnorderedSigh** [comment]
I'm hoping @Hurleveur will be available to review the tests. I could add quite wonderful features to conversations if they didn't have this bug.

File: source/PlayerInfo.cpp [hunk 5]
```diff
@@ -3443,11 +3467,23 @@ void PlayerInfo::Autosave() const
 
 
 
-void PlayerInfo::Save(const string &path) const
+void PlayerInfo::Save(const std::string &filePath) const
```
Comment (Hurleveur) [inline source/PlayerInfo.cpp:3470]: Why are you adding std::? Its not required and seems out of scope
We dont do it in the cpp since we include namespace std
```suggestion
void PlayerInfo::Save(const string &filePath) const
```

File: source/PlayerInfo.cpp [hunk 6]
```diff
@@ -3443,11 +3467,23 @@ void PlayerInfo::Autosave() const
...
 {
-	DataWriter out(path);
+	if(transactionSnapshot)
+	{
+		transactionSnapshot->SaveToPath(filePath);
+	}
```
Comment (Hurleveur) [inline source/PlayerInfo.cpp:3475]: Dont need these braces since its a contained instruction (here, within one line even)

File: tests/integration/config/plugins/integration-tests/data/tests/tests_save_load.txt [hunk 7]
```diff
@@ -1,3 +1,97 @@
...
+		call "Land On Ruin helper"
+		call "Trigger Spaceport Mission"
+		assert
+			"%TEST%: transient" == 1
+		# Re-load savegame mid-conversation.
+		input
+			key escape
+		call "Load First Savegame"
+		assert
+			"%TEST%: transient" == 0
```
Comment (Hurleveur) [inline tests/integration/config/plugins/integration-tests/data/tests/tests_save_load.txt:93]: Shouldnt we test that we can get that state back?
I assume that this means it was tested when we were landed before triggering the convo, right? So just try it again.

And then, reload *again* without saving to make sure it didnt auto save at the end of the convo? I think the code currently does that which is a bit weird if you ask me

File: source/PlayerInfo.cpp (hunk 6 above)
Comment (wjp) [reply source/PlayerInfo.cpp]: Done

**Comment by wjp** [comment]
One more question to answer is if we should save the recently added Global Conditions during conversations already, or postpone saving those too.

File: tests/integration/config/plugins/integration-tests/data/tests/tests_save_load.txt (hunk 7 above)
Comment (Hurleveur) [reply tests/integration/config/plugins/integration-tests/data/tests/tests_save_load.txt:93]: I mean eventually we could have an option where you choose when the game is saved

Comment (wjp) [reply tests/integration/config/plugins/integration-tests/data/tests/tests_save_load.txt:93]: I've added a section to the test going through the conversation again to see the mission was indeed triggered again after re-loading. Not entirely sure what you mean by the last bit about testing saving after the conversation though.

Comment (Hurleveur) [reply tests/integration/config/plugins/integration-tests/data/tests/tests_save_load.txt:93]: Well simply trying to save (happens when you exit, right?) once the convo is done, load that save and see if the condition is still set

Comment (wjp) [reply tests/integration/config/plugins/integration-tests/data/tests/tests_save_load.txt:93]: Ah I see, that makes sense. Added now.

**Comment by Hurleveur** [comment]
> One more question to answer is if we should save the recently added Global Conditions during conversations already, or postpone saving those too.

I dont think they should have special treatment here honestly, postpone them?

**Comment by Hecter94** [comment]
It looks like the feedback Hurleveur left was addressed, this seems to be fixing a fairly major bug so I'm hoping @warp-core can give it the final approval and merge if it looks good. 
Would be good to have this in for 0.10.0

**Comment by UnorderedSigh** [comment]
Players can corrupt their saves by accident if they save mid-conversation at the wrong point. Also, you can get free ships and money if you abuse the bug. It's a pretty bad bug, so I hope this fix will be merged into the 0.10.0 release.

**Comment by UnorderedSigh** [comment]
A github check is failing due to one of microsoft's servers being inaccessible.

**Comment by UnorderedSigh** [comment]
I just retested, and I can confirm the fix still works after master was merged into this branch.

**Review by Hurleveur — APPROVED** [review]
My feedback was indeed affressed, the integration test now fully makes sure this works
Looks good to me code wise.

**Comment by Zireael07** [comment]
Going by Reddit, some of the Canonical's servers are having issues today

**Review by UnorderedSigh — APPROVED** [review]

**Comment by warp-core** [comment]
While testing this, it seems like this makes it possible to skip an `on complete` action if you reload while the `on complete` conversation is active, so whatever else was in that action won't have been applied, but the mission will still be marked as complete and removed from the save file.

**Comment by UnorderedSigh** [comment]
I think this needs to be restricted to `on offer` conversations for it to work properly.

File: source/ConversationPanel.h [hunk 8]
```diff
@@ -107,6 +108,9 @@ template <class T>
 	// Reference to the player, to apply any changes to them.
 	PlayerInfo &player;
 
+  	// Should we use a PlayerInfo transaction to prevent save-load glitches?
```
Comment (warp-core) [inline source/ConversationPanel.h:111]:
```suggestion
	// Should we use a PlayerInfo transaction to prevent save-load glitches?
```

**Review by warp-core — APPROVED** [review]
